Modules/_sqlite/connection.c

```diff
@@ -396,43 +396,29 @@ static PyObject *
 pysqlite_connection_commit_impl(pysqlite_Connection *self)
 /*[clinic end generated code: output=3da45579e89407f2 input=39c12c04dda276a8]*/
 {
-    int rc;
-    sqlite3_stmt* statement;
-
     if (!pysqlite_check_thread(self) || !pysqlite_check_connection(self)) {
         return NULL;
     }
 
     if (!sqlite3_get_autocommit(self->db)) {
+        int rc;
 
         Py_BEGIN_ALLOW_THREADS
+        sqlite3_stmt *statement;
         rc = sqlite3_prepare_v2(self->db, "COMMIT", 7, &statement, NULL);
-        Py_END_ALLOW_THREADS
-        if (rc != SQLITE_OK) {
-            _pysqlite_seterror(self->state, self->db);
-            goto error;
+        if (rc == SQLITE_OK) {
+            (void)sqlite3_step(statement);
+            rc = sqlite3_finalize(statement);
         }
-
-        rc = pysqlite_step(statement);
-        if (rc != SQLITE_DONE) {
-            _pysqlite_seterror(self->state, self->db);
-        }
-
-        Py_BEGIN_ALLOW_THREADS
-        rc = sqlite3_finalize(statement);
```

Comment:
Seems that `sqlite3_finalize` was protected with the GIL previously and the check for `PyErr_Occurred` makes me thing that some callback or similar could end calling Python code. Could you confirm if this is indeed safe?

Reply:
Hm, `sqlite3_finalize` was called outside of the GIL previously, and it still is. Previously, we would release/acquire GIL before/after every SQLite API call; now we release, do all SQLite API calls, then acquire.

### Previously
1. allow threads (unlock GIL), prepare statement, disallow threads (lock GIL)
2. raise exception _and_ bail on error
3. unlock GIL, execute SQL statement (`sqlite3_step`), lock GIL
4. raise exception _and "fall through"_ on error
5. unlock GIL, finalise statement (release resources, etc.), lock GIL
6. raise exception, unless `sqlite3_step` already did, on error => this is the extra `PyErr_Occurred`

### Now
1. unlock GIL, create, execute, and finalise SQL statement, lock GIL
2. raise exception and bail on error

Reply:
Can `sqlite3_finalize` raise Python errors on its own?

Reply:
Not the way we currently use the SQLite API. If we'd added SQLite profile support, `sqlite3_finalize` (and other API's) would invoke our `_trace_callback()` C callback with argument `type` set to `SQLITE_TRACE_PROFILE`. We could add an `assert(type != SQLITE_TRACE_PROFILE)` and a comment about this to `_trace_callback()`.